docs/job-writing-guide.md

```diff
@@ -0,0 +1,750 @@
+
+```js
+const newState = {
+  ...state,
+};
+```
+
+Properties are declared in sequence, so you can spread an object and then
+declare more properties:
+
+```js
+const newState = {
+  ...state
+  data: {} // create a new data object but keep all other keys of state
+}
+```
+
+You can spread multiple objects, which again will apply in order. This example
+applies some default values, then overwrites with whatever is on state, then
+finally overwrites the data key.
+
+```js
+const newState = {
+  ...defaults,
+  ...state
+  data: {} // create a new data object but keep all other keys of state
+}
+```
+
+Spreading like this does not affect the original object (ie, in the example
+above, `defaults` and `state` are not changed), although remember that this is
+only a shallow clone, and non-primitive values use pointers, not copies.
+
+<details>
+<summary>What is a shallow clone?</summary>
+To shallow clone an object means to copy all the top-level keys and values of that object onto a new object.
+
+But this ONLY applies to top-level keys. And if a value contains an object,
+you're really just copying a _pointer_ to that object.
+
+```js
+const a = {
+  x: 1,
+  y: {
+    values: [1, 2, 3]
+  }
+};
+
+// declare b as a shallow clone of a
+const b = {
+  ... a
+}
+
+b.x = 2; // a.x is unchanged
+b.y.values = []; // a.y.values is changed
+b.y = 20' // a.y is unchanged
+```
+
+A deep clone means that all properties in the whole object tree are cloned.
+
+</details>
+
+## Compilation
+
+The code you write isn't technically executable Javascript. You can't just run
+it through node.js. It needs to be transformed or compliled into portable
+vanilla JS code.
+
+:::Warning This is advanced stuff more focused at JavaScript developers and the
+technically curious. These docs are not intended to be complete - just a nudge
+in the right direction to help understand how jobs work. :::
+
+The major differences between openfn code and Javascript are:
+
+- The top level functions in the code are executed synchronously (in sequence),
+  even if they contain asynchronous code
+- OpenFn code does not contain import statements (although technically it can).
+  These are compiled in.
+- Compiled code is a JavaScript ESM module which default-exports an array of
+  async functions. These functions are imported and executed by the runtime.
+
+It shouldn't be necessary to understand compilation in detail, but you should be
+aware that the code you write is not the code you run.
+
+If you're a Javascript developer, understanding some of these changes might help
+you better understand how OpenFn works. Using the CLI, you can run
+`openfn compile path/to/job.ja -a <adaptor-name>` to see compiled code.
+
+Here's an example of how a simple job looks in compilation:
+
+This job:
+
+````
+get('/patients')
+```js
+Compiles to this Javascript module:
+```js
+import { get } from "@openfn/language-http";
+export * from "@openfn/language-http";
+export default [get('/patients')];
+````
```

Comment:
@josephjclark at the end, I think we should direct folks to where to go next/learn more... maybe to the main [job docs page](https://docs.openfn.org/documentation/build/steps/jobs)? 
And how about a call to action, such as asking folks to post questions on Community if they get stuck?